File: modules/swagger-jaxrs/src/main/scala/com/wordnik/swagger/jaxrs/config/BeanConfig.scala
package com.wordnik.swagger.jaxrs.config

import com.wordnik.swagger.annotations.Api
import com.wordnik.swagger.core.{ SwaggerContext, SwaggerSpec }
import com.wordnik.swagger.model.ApiInfo
import com.wordnik.swagger.jaxrs.reader.DefaultJaxrsApiReader
import com.wordnik.swagger.config._
import com.wordnik.swagger.reader._
import com.wordnik.swagger.core.filter._
import org.reflections.Reflections
import org.reflections.scanners.{ SubTypesScanner, TypeAnnotationsScanner }
import org.reflections.util.{ ClasspathHelper, ConfigurationBuilder}
import org.slf4j.LoggerFactory
import javax.servlet.ServletConfig
import javax.ws.rs.core.Application
import scala.collection.JavaConverters._
import scala.beans.BeanProperty
import javax.ws.rs.Path

class BeanConfig extends JaxrsScanner {
  private val LOGGER = LoggerFactory.getLogger(classOf[BeanConfig])
  
  ConfigFactory.config = new SwaggerConfig
  ClassReaders.reader = Some(new DefaultJaxrsApiReader)

  @BeanProperty var resourcePackage: String = _
  @BeanProperty var title: String = _
  @BeanProperty var description: String = _ 
  @BeanProperty var termsOfServiceUrl: String = _ 
  @BeanProperty var contact: String = _ 
  @BeanProperty var license: String = _ 
  @BeanProperty var licenseUrl: String = _
  @BeanProperty var filterClass: String = _

  def setScan(shouldScan: Boolean) = {
    if(title != null || description != null || termsOfServiceUrl != null || contact != null || license != null || licenseUrl != null) {
      ConfigFactory.config.info = Some(ApiInfo(title, description, termsOfServiceUrl, contact, license, licenseUrl))
    }
    if(filterClass != null) {
      try {
        FilterFactory.setFilter(SwaggerContext.loadClass(filterClass).newInstance.asInstanceOf[SwaggerSpecFilter])
      }
      catch {
        case e: Exception => {
          LOGGER.error("failed to load filter", e)
        }
      }
    }
    ScannerFactory.scanner = Some(this)
  }

  def getScan() = true

  def classesFromContext(app: Application, sc: ServletConfig) : List[Class[_]] = {
    val config = new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage(resourcePackage)).setScanners(
      new TypeAnnotationsScanner(), new SubTypesScanner())
    var relections = new Reflections(config)
    (relections.getTypesAnnotatedWith(classOf[Api]).asScala ++ 
      relections.getTypesAnnotatedWith(classOf[Path]).asScala ).toList
  }

  def setApiReader(reader: String) = {
    try{
      val cls = SwaggerContext.loadClass(reader).newInstance.asInstanceOf[ClassReader]
      ClassReaders.reader = Some(cls)
    }
    catch {
      case e: Exception => LOGGER.error("failed to load reader class %s".format(reader))
    }
  }


  def getApiReader(): String = ClassReaders.reader match {
    case Some(cls) => cls.getClass.getName
    case _ => ""
  }

  def setVersion(apiVersion: String) = ConfigFactory.config.apiVersion = apiVersion
  def getVersion(): String = ConfigFactory.config.apiVersion

  def setBasePath(basePath: String) = ConfigFactory.config.basePath = basePath
  def getBasePath(): String = ConfigFactory.config.basePath
}

File: modules/swagger-jaxrs/src/main/java/com/wordnik/swagger/jaxrs/config/BeanConfig.java
package com.wordnik.swagger.jaxrs.config;

import com.wordnik.swagger.models.*;
import com.wordnik.swagger.jaxrs.Reader;
import com.wordnik.swagger.config.*;

import com.wordnik.swagger.annotations.Api;
import com.wordnik.swagger.core.filter.*;
import com.wordnik.swagger.config.FilterFactory;

import org.reflections.Reflections;
import org.reflections.scanners.ResourcesScanner;
import org.reflections.scanners.SubTypesScanner;
import org.reflections.scanners.TypeAnnotationsScanner;
import org.reflections.util.ClasspathHelper;
import org.reflections.util.ConfigurationBuilder;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Set;
import java.util.HashSet;

public class BeanConfig extends AbstractScanner implements Scanner, SwaggerConfig {
  Logger LOGGER = LoggerFactory.getLogger(BeanConfig.class);

  Reader reader = new Reader(new Swagger());

  String resourcePackage;
  String [] schemes;
  String title;
  String version;
  String description;
  String termsOfServiceUrl;
  String contact;
  String license;
  String licenseUrl;
  String filterClass;

  Info info;
  String host;
  String basePath;

  public String getResourcePackage() {
    return this.resourcePackage;
  }
  public void setResourcePackage(String resourcePackage) {
    this.resourcePackage = resourcePackage;
  }

  public String[] getSchemes() {
    return schemes;
  }
  public void setSchemes(String[] schemes) {
    this.schemes = schemes;
  }

  public String getTitle() {
    return title;
  }
  public void setTitle(String title) {
    this.title = title;
  }

  public String getVersion() {
    return version;
  }
  public void setVersion(String version) {
    this.version = version;
  }

  public String getDescription() {
    return description;
  }
  public void setDescription(String description) {
    this.description = description;
  }

  public String getTermsOfServiceUrl() {
    return termsOfServiceUrl;
  }
  public void setTermsOfServiceUrl(String termsOfServiceUrl) {
    this.termsOfServiceUrl = termsOfServiceUrl;
  }

  public String getContact() {
    return contact;
  }
  public void setContact(String contact) {
    this.contact = contact;
  }

  public String getLicense() {
    return license;
  }
  public void setLicense(String license) {
    this.license = license;
  }

  public String getLicenseUrl() {
    return licenseUrl;
  }
  public void setLicenseUrl(String licenseUrl) {
    this.licenseUrl = licenseUrl;
  }

  public Info getInfo() {
    return info;
  }
  public void setInfo(Info info) {
    this.info = info;
  } 

  public String getHost() {
    return host;
  }
  public void setHost(String host) {
    this.host = host;
  }

  public String getFilterClass() {
    return filterClass;
  }
  public void setFilterClass(String filterClass) {
    this.filterClass = filterClass;
  }  

  public String getBasePath() {
    return basePath;
  }
  public void setBasePath(String basePath) {
    if(!"".equals(basePath) && basePath != null) {
      if(!basePath.startsWith("/"))
        this.basePath = "/" + basePath;
      else
        this.basePath = basePath;
    }
  }

  public void setPrettyPrint(String prettyPrint) {
    if(prettyPrint != null)
      this.prettyPrint = Boolean.parseBoolean(prettyPrint);
  }

  public void setScan(boolean shouldScan) {
    Set<Class<?>> classes = classes();
    if(classes != null)
      reader.read(classes)
        .host(host)
        .basePath(basePath)
        .info(info);
    ScannerFactory.setScanner(this);
  }

  public boolean getScan() {
    return true;
  }

  public Set<Class<?>> classes() {
    ConfigurationBuilder config = new ConfigurationBuilder();
    Set<String> acceptablePackages = new HashSet<String>();

    if(resourcePackage != "") {
      String[] parts = resourcePackage.split(",");
      for(String pkg : parts) {
        if(!"".equals(pkg)) {
          acceptablePackages.add(pkg);
          config.addUrls(ClasspathHelper.forPackage(pkg));
        }
      }
    }

    config.setScanners(new ResourcesScanner(), new TypeAnnotationsScanner(), new SubTypesScanner());

    this.info = new Info()
      .description(description)
      .title(title)
      .version(version)
      .termsOfService(termsOfServiceUrl);

    if(contact != null)
      this.info.contact(new Contact()
        .name(contact));
    if(license != null && licenseUrl != null)
      this.info.license(new License()
        .name(license)
        .url(licenseUrl));
    if(schemes != null) {
      for(String scheme : schemes)
        reader.getSwagger().scheme(Scheme.forValue(scheme));
    }

    reader.getSwagger().setInfo(info);
    Set<Class<?>> classes = new Reflections(config).getTypesAnnotatedWith(Api.class);
    Set<Class<?>> output = new HashSet<Class<?>>();
    for(Class<?> cls : classes) {
      if(acceptablePackages.contains(cls.getPackage().getName()))
        output.add(cls);
    }
    return output;
  }

  public Swagger getSwagger() {
    return reader.getSwagger();
  }

  public Swagger configure(Swagger swagger) {
    if(schemes != null) {
      for(String scheme : schemes)
        swagger.scheme(Scheme.forValue(scheme));
    }
    if(filterClass != null) {
      try {
        SwaggerSpecFilter filter = (SwaggerSpecFilter) Class.forName(filterClass).newInstance();
        if(filter != null) {
          FilterFactory.setFilter(filter);
        }
      }
      catch (Exception e) {
        LOGGER.error("failed to load filter", e);
      }
    }
    return swagger.info(info)
      .host(host)
      .basePath(basePath);
  }
}
